rust/kernel/xarray.rs

```diff
@@ -0,0 +1,299 @@
+        // outlive the `XArray` itself, nor the Guard that holds the lock ensuring the value
+        // remains in the `XArray`.
+        unsafe { T::borrow(self.0 as _) }
+    }
+}
+
+// Convenience impl for `ForeignOwnable` types whose `Borrowed`
+// form implements Deref.
+impl<'a, T: ForeignOwnable> Deref for Guard<'a, T>
+where
+    T::Borrowed<'static>: Deref,
+{
+    type Target = <T::Borrowed<'static> as Deref>::Target;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: See the `borrow()` method. The dereferenced `T::Borrowed` value
+        // must share the same lifetime, so we can return a reference to it.
+        // TODO: Is this really sound?
+        unsafe { &*(T::borrow(self.0 as _).deref() as *const _) }
+    }
+}
+
+impl<'a, T: ForeignOwnable> Drop for Guard<'a, T> {
+    fn drop(&mut self) {
+        // SAFETY: The XArray we have a reference to owns the C xarray object.
+        unsafe { bindings::xa_unlock(self.1.get()) };
+    }
+}
+
+/// Represents a reserved slot in an `XArray`, which does not yet have a value but has an assigned
+/// index and may not be allocated by any other user. If the Reservation is dropped without
+/// being filled, the entry is marked as available again.
+///
+/// Users must ensure that reserved slots are not filled by other mechanisms, or otherwise their
+/// contents may be dropped and replaced (which will print a warning).
+pub struct Reservation<'a, T: ForeignOwnable>(&'a XArray<T>, usize, PhantomData<T>);
+
+impl<'a, T: ForeignOwnable> Reservation<'a, T> {
+    /// Store a value into the reserved slot.
+    pub fn store(self, value: T) -> Result<usize> {
+        if self.0.replace(self.1, value)?.is_some() {
+            crate::pr_err!("XArray: Reservation stored but the entry already had data!\n");
+            // Consider it a success anyway, not much we can do
+        }
+        let index = self.1;
+        core::mem::forget(self);
+        Ok(index)
+    }
+
+    /// Returns the index of this reservation.
+    pub fn index(&self) -> usize {
+        self.1
+    }
+}
+
+impl<'a, T: ForeignOwnable> Drop for Reservation<'a, T> {
+    fn drop(&mut self) {
+        if self.0.remove(self.1).is_some() {
+            crate::pr_err!("XArray: Reservation dropped but the entry was not empty!\n");
+        }
+    }
+}
+
+/// An array which efficiently maps sparse integer indices to owned objects.
+///
+/// This is similar to a `Vec<Option<T>>`, but more efficient when there are holes in the
+/// index space, and can be efficiently grown.
+///
+/// This structure is expected to often be used with an inner type that can either be efficiently
+/// cloned, such as an `Arc<T>`.
+pub struct XArray<T: ForeignOwnable> {
+    xa: Opaque<bindings::xarray>,
+    _p: PhantomData<T>,
+}
+
+impl<T: ForeignOwnable> XArray<T> {
+    /// Creates a new `XArray` with the given flags.
+    pub fn new(flags: Flags) -> Result<XArray<T>> {
```

Comment:
Could this constructor be made constant? It would be pretty useful for static mutexes, instead of using a lazy value like `LazyCell`.

Reply:
Unfortunately not. It calls the `xa_init_flags` foreign function. Const functions are only allowed to call other const functions defined in rust.

Reply:
Then the way to make it happen would be to either inline the `xa_init_flags` function or have another constructor that declares an `xarray` in place, though either ways still require a raw `spinlock_t`.

https://github.com/Rust-for-Linux/linux/blob/1dd685c414a7b9fdb3d23aca3aedae84f0b998ae/include/linux/xarray.h#L296-L301

Reply:
Wait, I have an idea.

```rust
    pub fn new(flags: Flags) -> Result<XArray<T>> {
        let xa = Opaque::uninit();

        // SAFETY: We have just created `xa`. This data structure does not require
        // pinning.
        unsafe {
            xa.get().write_bytes(0, 1);
            (*xa.get()).xa_flags = flags;
        }

        // INVARIANT: Initialize the `XArray` with a valid `xa`.
        XArray {
            xa,
            _p: PhantomData,
        }
    }
```

Reply:
`spin_lock_init` is more complicated then `write_bytes(0, 1)`, but I guess it's doable.

Reply:
I don't think I'm really comfortable with this, with all the spinlock variations involved (RT builds, lockdep, etc.)... I don't think the upstream maintainers of all that will want to duplicate all of that logic in Rust...

Reply:
Oh, I wasn't talking about upstreaming this, just it's doable, which can give us some ideas in the future how much we can gain if we make XA new `const`.